src/Data/MeasuredIn.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- Representation of values with an associated (free) unit of measure. Useful to
+-- disambiguate primitive types like 'Int' or 'String' which can be in different
+-- bases depending on the context.
+
+module Data.MeasuredIn
+    ( -- * Polymorphic MeasuredIn
+      MeasuredIn(..)
+
+      -- * Percentage ~ MeasuredIn "percent" Word
+    , MkPercentageError(..)
+    , mkPercentage
+    ) where
+
+import Prelude
+
+import Data.Aeson
+    ( FromJSON (..)
+    , ToJSON (..)
+    , Value (String)
+    , object
+    , withObject
+    , (.:)
+    , (.=)
+    )
+import Data.Aeson.Types
+    ( Parser )
+import Data.Proxy
+    ( Proxy (..) )
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( KnownSymbol, Symbol, symbolVal )
+
+import qualified Data.Text as T
+
+
+-- | Represents a value that has an associated unit of measure, based on some
+--   underlying type.
+--
+-- >>> newtype Amount = Amount (MeasuredIn "lovelace" Word32)
+newtype MeasuredIn (u :: Symbol) a = MeasuredIn a
+    deriving (Generic, Show, Eq)
+
+-- | Encode to JSON delegating the
+--
+-- >>> Aeson.encode $ MeasuredIn @"lovelace" 14
+-- {"unit":"lovelace","quantity":14}
+instance (KnownSymbol u, ToJSON a) => ToJSON (MeasuredIn u a) where
+    toJSON (MeasuredIn a) = object
+        [ "unit"     .= symbolVal (Proxy :: Proxy u)
+        , "quantity" .= toJSON a
+        ]
+
+instance (KnownSymbol u, FromJSON a) => FromJSON (MeasuredIn u a) where
+    parseJSON = withObject "MeasuredIn" $ \o -> do
+        verifyUnit (Proxy :: Proxy u) =<< o .: "unit"
+        MeasuredIn <$> o .: "quantity"
+
+
+{-------------------------------------------------------------------------------
+                                Percentage
+-------------------------------------------------------------------------------}
+
+deriving newtype instance Integral i => Ord (MeasuredIn "percent" i)
+
+instance Integral i => Bounded (MeasuredIn "percent" i) where
+    minBound = MeasuredIn 0
+    maxBound = MeasuredIn 100
+
+instance {-# OVERLAPS #-} (Integral i, FromJSON i) => FromJSON (MeasuredIn "percent" i) where
```

Comment:
This line is overly long, according to our coding standards. Why not wrap this as follows:

```hs
instance {-# OVERLAPS #-} (Integral i, FromJSON i) =>
    FromJSON (MeasuredIn "percent" i) where
```